solver/llbsolver/history.go

```diff
@@ -17,24 +30,117 @@ const (
 type HistoryQueueOpt struct {
 	DB           *bolt.DB
 	LeaseManager leases.Manager
+	ContentStore content.Store
+	CleanConfig  *config.HistoryConfig
 }
 
 type HistoryQueue struct {
 	mu       sync.Mutex
 	initOnce sync.Once
 	HistoryQueueOpt
-	ps     *pubsub[*controlapi.BuildHistoryEvent]
-	active map[string]*controlapi.BuildHistoryRecord
+	ps      *pubsub[*controlapi.BuildHistoryEvent]
+	active  map[string]*controlapi.BuildHistoryRecord
+	refs    map[string]int
+	deleted map[string]struct{}
 }
 
 func NewHistoryQueue(opt HistoryQueueOpt) *HistoryQueue {
-	return &HistoryQueue{
+	if opt.CleanConfig == nil {
+		opt.CleanConfig = &config.HistoryConfig{
+			MaxAge:     int64((48 * time.Hour).Seconds()),
+			MaxEntries: 50,
+		}
+	}
+	h := &HistoryQueue{
 		HistoryQueueOpt: opt,
 		ps: &pubsub[*controlapi.BuildHistoryEvent]{
 			m: map[*channel[*controlapi.BuildHistoryEvent]]struct{}{},
 		},
-		active: map[string]*controlapi.BuildHistoryRecord{},
+		active:  map[string]*controlapi.BuildHistoryRecord{},
+		refs:    map[string]int{},
+		deleted: map[string]struct{}{},
+	}
+
+	go func() {
+		for {
+			h.gc()
+			time.Sleep(120 * time.Second)
+		}
+	}()
+
+	return h
+}
+
+func (h *HistoryQueue) gc() error {
+	var records []*controlapi.BuildHistoryRecord
+
+	if err := h.DB.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(recordsBucket))
+		if b == nil {
+			return nil
+		}
+		return b.ForEach(func(key, dt []byte) error {
+			var br controlapi.BuildHistoryRecord
+			if err := br.Unmarshal(dt); err != nil {
+				return errors.Wrapf(err, "failed to unmarshal build record %s", key)
+			}
+			if br.Pinned {
+				return nil
+			}
+			records = append(records, &br)
+			return nil
+		})
+	}); err != nil {
+		return err
+	}
+
+	if len(records) < int(h.CleanConfig.MaxEntries) {
+		return nil
+	}
+
+	sort.Slice(records, func(i, j int) bool {
+		return records[i].CompletedAt.Before(*records[j].CompletedAt)
+	})
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	now := time.Now()
+	for _, r := range records[h.CleanConfig.MaxEntries:] {
+		if now.Add(time.Duration(h.CleanConfig.MaxAge) * -time.Second).After(*r.CompletedAt) {
+			if err := h.delete(r.Ref, false); err != nil {
+				return err
+			}
+		}
 	}
+
+	return nil
+}
+
+func (h *HistoryQueue) delete(ref string, sync bool) error {
+	if _, ok := h.refs[ref]; ok {
+		h.deleted[ref] = struct{}{}
+		return nil
+	}
+	if err := h.DB.Update(func(tx *bolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists([]byte(recordsBucket))
+		if err != nil {
+			return err
+		}
+		err1 := b.Delete([]byte(ref))
```

Comment:
Do we need to create the bucket if it doesn't exist here, just to try and perform a deletion?